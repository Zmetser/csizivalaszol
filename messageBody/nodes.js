/**
 * @flow
 *
 * All the usable nodes
 */

export type InlineStyle = 'Italic' | 'Bold' | 'Struck' | 'Small' | 'Code' | 'Subscript' | 'Superscript' | 'Underline' | 'Normal'

const TextNode = (value: string, styles: Array<InlineStyle>) => ({
  type: 'Text',
  value,
  styles
})

const LineBreak = () => ({ type: 'LineBreak' })

const Link = (value: Array<TextNode>, href: string, target?: ?string) => ({
  type: 'Link',
  value,
  href,
  target
})

export type InlineNode = TextNode | LineBreak | Link;

type ImageAttributes = { src: string, alt: string, width: number, height: number }
const Image = ({ src, alt, width, height }: ImageAttributes) => ({
  type: 'Image',
  src,
  alt,
  width,
  height
})

const Paragraph = (value: Array<InlineNode>) => ({
  type: 'Paragraph',
  value
})

module.exports = {
  TextNode,
  Paragraph,
  LineBreak,
  Link,
  Image
}
