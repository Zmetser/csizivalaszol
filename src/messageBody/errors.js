export class UnknownNodeError extends Error {
  constructor (node: BlockNode) {
    super(node.type)
  }
}
UnknownNodeError.prototype.name = 'UnknownNodeError'
