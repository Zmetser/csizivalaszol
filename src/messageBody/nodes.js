/**
 * All the usable nodes
 */

import type {
  ImageAttributes,
  ImageNode,
  InlineNode,
  InlineStyle,
  LineBreakNode,
  LinkNode,
  ParagraphNode,
  TextNode,
  ListType,
  ListItem,
  ListNode
} from './types'

export const createTextNode = (value: string, styles: Array<InlineStyle>): TextNode => ({
  type: 'Text',
  value,
  styles
})

export const createLineBreak = (): LineBreakNode => ({ type: 'LineBreak' })

export const createLink = (value: Array<TextNode>, href: string, target?: ?string): LinkNode => ({
  type: 'Link',
  value,
  href,
  target
})

export const createImage = ({ src, alt, width, height }: ImageAttributes): ImageNode => ({
  type: 'Image',
  src,
  alt,
  width,
  height
})

export const createParagraph = (value: Array<InlineNode>): ParagraphNode => ({
  type: 'Paragraph',
  value
})

export function createList<T: ListType> (items: Array<ListItem>, listType: T): ListNode<T> {
  return {
    type: 'List',
    listType,
    items
  }
}
