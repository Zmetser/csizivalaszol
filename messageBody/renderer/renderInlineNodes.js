/**
 * @flow
 */

import * as React from 'react'
import { UnknownNodeError } from '../errors'

import type { InlineNode } from '../types'

import {
  Text,
  Link
} from '../components'

function createInlineNode (node: InlineNode, index): React.Element<*> {
  switch (node.type) {
    case 'Text': return Text(node, index)
    case 'Link': return <Link {...node} key={index} />
    default: throw new UnknownNodeError(node)
  }
}
export default function renderInlineNodes (nodes: Array<InlineNode>): Array<React.Element<*>> {
  return nodes.map(createInlineNode)
}
