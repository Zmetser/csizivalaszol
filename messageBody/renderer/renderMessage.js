/**
 * @flow
 */

import * as React from 'react'
import { Paragraph } from '../components'

import type { BlockNode } from '../types.js'
import { UnknownNodeError } from '../errors'

export default function render (message: Array<BlockNode>): Array<React.Element<*>> {
  return message.map((node, index): React.Element<*> => {
    switch (node.type) {
      case 'Paragraph': return Paragraph({ value: node.value }, index)
      default: throw new UnknownNodeError(node)
    }
  })
}
