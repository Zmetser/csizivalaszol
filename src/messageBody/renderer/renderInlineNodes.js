/**
 * @flow
 */

import * as React from 'react'
import { UnknownNodeError } from '../errors'

import type { InlineNode } from '../types'

import {
  Text,
  Link,
  LineBreak
} from '../components'

function createInlineNode (node: InlineNode, index): React.Element<*> {
  switch (node.type) {
    case 'Text': return <Text node={node} key={index} />
    case 'Link': return <Link node={node} key={index} />
    case 'LineBreak': return <LineBreak key={index} />
    default: throw new UnknownNodeError(node)
  }
}
export default function renderInlineNodes (nodes: Array<InlineNode>): Array<React.Element<*>> {
  return nodes.map(createInlineNode)
}
