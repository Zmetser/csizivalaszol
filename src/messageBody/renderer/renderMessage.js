import * as React from 'react'
import {
  Paragraph,
  List
} from '../components'

import type { BlockNode } from '../types.js'
import { UnknownNodeError } from '../errors'

export default function render (message: Array<BlockNode>): Array<React.Element<*>> {
  return message.map((node, index): React.Element<*> => {
    switch (node.type) {
      case 'Paragraph': return <Paragraph node={node} key={index} />
      case 'List': return <List node={node} key={index} />
      default: throw new UnknownNodeError(node)
    }
  })
}
