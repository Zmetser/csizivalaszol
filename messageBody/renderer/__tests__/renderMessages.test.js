/**
 * @flow
 */

import * as React from 'react'
import renderer from 'react-test-renderer'

import { UnknownNodeError } from '../../errors'
import renderMessage from '../renderMessage'

import { createParagraph, createTextNode } from '../../nodes'

describe('Render a single message', () => {
  test('with an unknown element, should throw UnknownNodeError', () => {
    // $FlowSkipForTesting: Testing what happens when a node is not supported.
    const unknownNode = [{'type': 'Foo'}]
    expect(() => renderMessage(unknownNode)).toThrowError(UnknownNodeError)
  })

  test('with a single paragraph', () => {
    const p1 = createParagraph([createTextNode('foo', [])])
    const element = renderer.create(
      <article>{renderMessage([p1])}</article>
    )
    expect(element.toJSON()).toMatchSnapshot()
  })

  test('with multiple paragraphs', () => {
    const p1 = createParagraph([createTextNode('foo', [])])
    const p2 = createParagraph([createTextNode('bar', [])])
    const element = renderer.create(
      <article>{renderMessage([p1, p2])}</article>
    )
    expect(element.toJSON()).toMatchSnapshot()
  })
})
