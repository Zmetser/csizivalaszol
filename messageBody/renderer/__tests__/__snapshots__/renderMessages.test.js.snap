// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a single message with a single paragraph 1`] = `
<article>
  <p>
    foo
  </p>
</article>
`;

exports[`Render a single message with multiple paragraphs 1`] = `
<article>
  <p>
    foo
  </p>
  <p>
    bar
  </p>
</article>
`;
